### PR TITLE
new ansible_parent_block_name for current task

### LINKED ISSUE #33581: Name of task containing block is not printed to stdout
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
Ansible Blocks

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [~/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]
```

##### OS / ENVIRONMENT
Red Hat Enterprise Linux Server release 7.4 (Maipo) (master and target)

##### SUMMARY
When a block is included in a named task the stdout when running the playbook does not include the name of the task containing the block.

##### STEPS TO REPRODUCE
```
- name: apply-custom-error-pages
  block:
    - copy:
        src=files/{{ env }}/404.jsp
        dest={{ liferay_webapps_portal }}
      notify:
        - apply-puppet-config
    - copy:
        src=files/{{ env }}/500.html
        dest={{ liferay_webapps_portal }}
      notify:
        - apply-puppet-config
```

##### EXPECTED RESULTS
The task name is echoed to stdout, perhaps with output like follows:
```
TASK [base-liferay : apply-custom-error-pages : copy] *******************************************************************************************************************************************************************************
changed: [hostname]

TASK [base-liferay : apply-custom-error-pages : copy] *******************************************************************************************************************************************************************************
changed: [hostname]
```
##### ACTUAL RESULTS
The task name is not echoed to stdout

```
TASK [base-liferay : copy] *******************************************************************************************************************************************************************************
changed: [hostname]

TASK [base-liferay : copy] *******************************************************************************************************************************************************************************
changed: [hostname]
```

### PR DESCRIPTION
**Pull request by bcoca**
So you can add to the task's own templating the name of the block that encapsulates it.

fixes #33581

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: test/integration/targets/special_vars/tasks/main.yml [hunk 1]
```diff
@@ -98,3 +98,16 @@
 - name: ansible_parent_role_names - test functionality by importing another role
   import_role:
     name: include_parent_role_vars
+
+- name: test block name
+  block:
+    - assert:
+        that:
+          - ansible_parent_block_name is defined
```
Comment (webknjaz) [inline test/integration/targets/special_vars/tasks/main.yml:106]:
```console
10:57     "msg": "The conditional check 'ansible_parent_block_name is None' failed. The error was: template error while templating string: Could not load \"None\": 'None'. String: {% if ansible_parent_block_name is None %} True {% else %} False {% endif %}. Could not load \"None\": 'None'"
```
*(https://dev.azure.com/ansible/ansible/_build/results?buildId=115759&view=logs&j=7afe5636-5efd-54c7-69c1-47871cffbb82&t=f3a51600-bef2-53db-6114-e5030a556751&l=15881)*

File: lib/ansible/playbook/task.py [hunk 2]
```diff
@@ -455,6 +455,15 @@ def set_loader(self, loader):
...
             self._parent.set_loader(loader)
 
+    def get_parent_block_name(self):
+
+        name = ''
+        parent_block = self.get_first_parent_block(static=True)
+        if parent_block:
+            name = getattr(parent_block, 'name', '')
+
+        return name
```
Comment (webknjaz) [inline lib/ansible/playbook/task.py]: Since the `getattr()` call has a fallback, I don't think that conditional is necessary:
```suggestion
        parent_block = self.get_first_parent_block(static=True)
        return getattr(parent_block, 'name', '')
```
Am I missing anything?

File: lib/ansible/playbook/task.py [hunk 3]
```diff
@@ -508,6 +517,21 @@ def get_first_parent_include(self):
...
+                if static:
+                    if getattr(self._parent, 'statically_loaded', True):
+                        parent = self._parent
+                    else:
+                        parent = self._parent.get_first_parent_block()
+                else:
+                    parent = self._parent
+            else:
+                parent = self._parent.get_first_parent_block()
+        return parent
```
Comment (webknjaz) [inline lib/ansible/playbook/task.py:534]: This seems suspiciously nested. How about flattening it a little?
```suggestion
        if not self._parent:
            return None

        if not isinstance(self._parent, Block):
            return self._parent.get_first_parent_block()

        if not static:
            return self._parent

        if getattr(self._parent, 'statically_loaded', True):
            return self._parent

        return self._parent.get_first_parent_block()
```

File: lib/ansible/playbook/task.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/playbook/task.py]: nope, i used to have more code there, but kept if after removing

File: lib/ansible/playbook/task.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/playbook/task.py:534]: i prefer aclear return and decision logic, having multiple returns can end up with different/incomplete returns.

File: changelogs/fragments/block_name.yml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - new constant 'ansible_parent_block_name' contains the 'name' of the current explicit block a task is in, it won't work with implicit/generated blocks.
```
Comment (mkrizek) [inline changelogs/fragments/block_name.yml]: I would probably not mention implicit blocks as it is an implementation detail.

**Review by mkrizek — APPROVED** [review]

File: lib/ansible/playbook/task.py (hunk 3 above)
Comment (webknjaz) [reply lib/ansible/playbook/task.py:534]: FWIW, highly nested branches are hard to follow and easier to make mistakes in.
